feat: Migrate from @sentry/apm to @sentry/tracing

AbhiPrasad opened the pull request:
Switch all references from `@sentry/apm` to `@sentry/tracing`.

Also add an upgrade guide detailing the changes.

AbhiPrasad commented on src/includes/performance-monitoring/configuration/javascript.mdx:
````diff
@@ -130,23 +131,33 @@ Sentry.init({
+```javascript
+const transaction = Sentry.startTransaction({ name: "test-transaction" });
+
+Sentry.configureScope(scope => scope.setSpan(transaction));
````
This section is new - added as I figured this would be a common concern to bring up.

rhcarvalho replied:
Hmm, I don't like sneaking this in "Migrate to @sentry/tracing" PR.

We could as well decide that startTransaction always updates the scope, making it less of a burden to associate errors and transactions.

lobsterkatie commented on src/docs/product/performance/apm-to-tracing.mdx:
```diff
@@ -0,0 +1,101 @@
+
+## Migrating from @sentry/apm to @sentry/tracing
+
+For most cases, just replacing your `@sentry/tracing` import is enough to switch packages.
```
```suggestion
For most cases, just replacing your `@sentry/apm` import is enough to switch packages.
```

lobsterkatie commented on src/docs/product/performance/apm-to-tracing.mdx:
````diff
@@ -0,0 +1,101 @@
+ });
+```
+
+If you were using the `beforeNavigate` option from the `Tracing` integration, the API for the `beforeNavigate` option in `BrowserTracing` has changed slightly. Instead of passing in a location and returning a string representing transaction name, `beforeNavigate` now accepts a transaction context and is expected to return transaction context. You can now add extra tags or change the op based on d ...
````
This will need to change once we sunset `beforeNavigate` in favor of `tracesSampler`. Do we want to hold off publishing this until that happens?

AbhiPrasad replied:
This is a good point. I will hold off until we make a decision in grooming tomorrow.

AbhiPrasad replied:
Update after grooming: have decided to document `beforeNavigate`, but not to mention filtering.

lobsterkatie commented on src/includes/performance-monitoring/configuration/javascript.mdx:
```diff
@@ -1,41 +1,42 @@
-import * as Sentry from '@sentry/browser';
-import { Integrations as ApmIntegrations } from '@sentry/apm';
+import * as Sentry from "@sentry/browser";
+import { Integrations } from "@sentry/tracing";
```
Did you change it to just `Integrations` for simplicity? It had been `ApmIntegrations` in order that people be able to import the integrations package as `Integrations` also. You might consider `TracingIntegrations`.

dcramer replied:
could we just improve the exports here?

```
import { Tracing } from "..."
```

What else would this even expose? Why do we need it to be contained in Integrations?

lobsterkatie replied:
> could we just improve the exports here?

I eventually came to the same idea as I read on, so I'm upgrading my "we could" below to a "I think we should." (Provided it doesn't break anything, which I don't believe it would.)

AbhiPrasad replied:
`import { Integrations } from "@sentry/tracing";`

`Integrations` is an object containing `{ Express, BrowserTracing }`

We could separate this yes, I just chose to make the behaviour consistent with `@sentry/apm`. It wouldn't break anything to change the imports, we would just have to update the upgrade process.

AbhiPrasad replied:
After revising this, I decided to not change the imports. In the CDN bundle, we keep everything together in `Sentry.Integrations`, hence why we chose to keep it the same way in the npm package. I think we should for now keep the same behaviour in `@sentry/apm` and we can revisit integrations in general in v6.

lobsterkatie commented on src/includes/performance-monitoring/configuration/javascript.mdx:
````diff
@@ -199,34 +210,37 @@ When Sentry captures transactions, they are assigned a transaction name. This na
 ```
 
-For browser JavaScript applications using the `Tracing` integration, the `beforeNavigate` option  can be used to better group navigation/pageload transactions together based on URL.
+For browser JavaScript applications using the `BrowserTracing` integration, the `beforeNavigate` option can be used to better group navigation/pageload transactions together based on URL.
````
@HazAT - this is a use of `beforeNavigate` which isn't covered by `tracesSampler`. We may need to rethink the idea of deprecating the former in favor of the latter.

lobsterkatie commented on src/includes/performance-monitoring/configuration/javascript.mdx:
````diff
@@ -199,34 +210,37 @@ When Sentry captures transactions, they are assigned a transaction name. This na
 ```
 
-For browser JavaScript applications using the `Tracing` integration, the `beforeNavigate` option  can be used to better group navigation/pageload transactions together based on URL.
+For browser JavaScript applications using the `BrowserTracing` integration, the `beforeNavigate` option can be used to better group navigation/pageload transactions together based on URL.
````
Could we roll this example into the nearly identical one above, by just mentioning grouping there? Are there any other differences between that example and this one?

AbhiPrasad replied:
There are no differences, we can combine the examples together yes.

lobsterkatie commented on src/includes/performance-monitoring/configuration/javascript.mdx:
```diff
@@ -235,15 +249,17 @@ Sentry.init({
 **Retrieving a Transaction**
 
-In cases where you want to attach Spans to an already ongoing Transaction you can use `Sentry.getCurrentHub().getScope().getTransaction()`. This function will return a `Transaction` in case there is a running Transaction otherwise it returns `undefined`. If you are using our Tracing integration by default we attach the Transaction to the Scope. So you could do something like this:
+In cases where you want to attach Spans to an already ongoing Transaction you can use `Sentry.getCurrentHub().getScope().getTransaction()`. This function will return a `Transaction` in case there is a running Transaction on the scope, otherwise it returns `undefined`. If you are using our BrowserTracing integration by default we attach the Transaction to the Scope. So you could do something like  ...
```
```suggestion
In cases where you want to attach Spans to an already ongoing Transaction you can use `Sentry.getCurrentHub().getScope().getTransaction()`. This function will return a `Transaction` in case there is a running Transaction on the scope, otherwise it returns `undefined`. If you are using our BrowserTracing integration, by default we attach the Transaction to the Scope. So you could do something like this:
```

Also - do we attach all transactions to the scope automatically when using the `BrowserTracing` integration, or just the `pageload` and `navigation` ones?

AbhiPrasad replied:
We only attach `pageload` and `navigation` transaction to the scope. I can mention that.

lobsterkatie commented on src/includes/performance-monitoring/configuration/koa.mdx:
```diff
@@ -85,8 +91,8 @@ app.use(tracingMiddleWare);
 app.on("error", (err, ctx) => {
-  Sentry.withScope(function (scope) {
-    scope.addEventProcessor(function (event) {
+  Sentry.withScope(function(scope) {
```
Is there a reason we're doing this with anonymous functions whereas we did the one above with arrow functions?

AbhiPrasad replied:
No reason. I think this documentation was community contributed, but I can switch to using arrow functions.

lobsterkatie commented on src/includes/performance-monitoring/configuration/react.mdx:
```diff
@@ -35,7 +41,7 @@ ReactDOM.render(<App />, rootNode);
 **`withProfiler` Higher-Order Component**
 
-`@sentry/react` exports a `withProfiler` higher-order component that leverages the `@sentry/apm` Tracing integration to add React related spans to transactions. If the Tracing integration is not enabled, the Profiler will not work.
+`@sentry/react` exports a `withProfiler` higher-order component that leverages the `@sentry/tracing` add React related spans to transactions.
```
```suggestion
`@sentry/react` exports a `withProfiler` higher-order component that leverages the `@sentry/tracing` to add React-related spans to transactions.
```

lobsterkatie commented on src/includes/performance-monitoring/configuration/vue.mdx:
```diff
@@ -1,49 +1,54 @@
-When Sentry encounters a component named `root`, which is a top-level Vue instance (as in `new Vue({})`), we use our AM Tracing integration,
-and create a new activity named `Vue Application Render`. Once the activity has been created, it will wait until all of its child components render, and there aren't new rendering events triggered within the configured `timeout`, before marking the activity as completed.
+When Sentry encounters a component named `root`, which is a top-level Vue instance (as in `new Vue({})`), we use our AM BrowserTracing integration,
+and create a new span named `Vue Application Render`. Once the span has been created, it will wait until all of its child components render, and there aren't new rendering events triggered within the configured `timeout`, before marking the span as finished.
```
Spans don't have timeouts, do they? Is this actually a transaction?

AbhiPrasad replied:
Ah this is based on `idleTransaction`. Maybe it is better just to remove this paragraph.

lobsterkatie commented on src/includes/performance-monitoring/configuration/vue.mdx:
```diff
@@ -1,49 +1,54 @@
 import * as Sentry from "@sentry/browser";
 import { Vue as VueIntegration } from "@sentry/integrations";
-import { Integrations } from "@sentry/apm";
+import { Integrations } from "@sentry/tracing";
```
If we're importing the entire tracing package here so that the `Sentry.init()` can look the same as it does everywhere else, perhaps we should do the same for the Vue integration? It might not be a bad idea to have at least one example where we're gracefully handling the `Integrations` naming conflict.

(Alternatively, we could handle every instance of this the other way, and only ever import the `BrowserTracing` integration instead of the whole package, which would make the question of naming conflicts moot.)

AbhiPrasad replied:
> only ever import the BrowserTracing integration instead of the whole package

We can make this change.

lobsterkatie commented on src/includes/performance-monitoring/configuration/vue.mdx:
```diff
@@ -1,49 +1,54 @@
 The described instrumentation functionality will give you very high-level information about the rendering performance of the Vue instance. However, the integration can also provide more fine-grained details about what actually happened during a specific activity.
 To do that, you need to specify which components to track and what hooks to listen to (you can find a list of all available hooks [here](https://vuejs.org/v2/api/#Options-Lifecycle-Hooks)). You can also turn on tracking for all the components. However, it may be rather noisy if your app consists of hundreds of components. We encourage being more specific. If you don't provide hooks, Sentry will t ...
 
 Note that we don't use `before` and `-ed` pairs for hooks, and you should provide a simple verb instead. For example, `update` is correct. `beforeUpdate` and `updated` are incorrect.
```
```suggestion
Note that when specifying hooks we use the simple verb rather than `before` and `-ed` pairs. For example, `destroy` is correct. `beforeDestroy` and `destroyed` are incorrect.
```

NB that I'm only pattern matching - I don't know for sure if `destroy` has a `before` and an `-ed`. I just thought that if possible it might be nice to have one that isn't the default since they already have examples of us correctly formatting the default ones.

AbhiPrasad replied:
@kamilogorek will know.

kamilogorek replied:
They all do, except `activate`. That's why we use one simple verbs instead. Here's the full mapping list - https://github.com/getsentry/sentry-javascript/blob/5ed313546dd687b637dfb20a3a28a7823278d21a/packages/integrations/src/vue.ts#L116-L123

lobsterkatie commented on src/includes/performance-monitoring/configuration/vue.mdx:
```diff
@@ -1,49 +1,54 @@
+Once you've configured the BrowserTracing integration, move on to configuring the Vue integration itself.
 Sentry built the new tracing capabilities into the original Vue error handler integrations, so there is no need to add any new packages. You only need to provide an appropriate configuration.
 
 The most basic configuration for tracing your Vue app, which would track only the top-level component, looks like this:
```
```suggestion
The most basic configuration for tracing your Vue app, which would track only the `mount` and `update` hooks for the top-level component, looks like this:
```

lobsterkatie commented on src/includes/performance-monitoring/configuration/vue.mdx:
```diff
@@ -78,10 +89,16 @@ new VueIntegration({
+      "RwvArticleList",
+      "Pagination",
+    ],
+    hooks: ["mount", "update", "destroy"],
```
IDK if a comment in the code here is the right place - perhaps it belongs in the explanatory text above - but it might be nice to mention that specifying the hooks yourself _replaces_ rather than adds to the default list, so `'mount'` and `'update'` need to be in there if they're still going to be tracked.

AbhiPrasad replied:
This is a good point, but something we can address in a future PR.

rhcarvalho commented on src/components/jsCdnTag.js:
```diff
@@ -19,7 +19,7 @@ export default ({ apm = false }) => {
   }
 
-  const packageName = apm ? "bundle.apm.min.js" : "bundle.min.js";
+  const packageName = tracing ? "bundle.tracing.min.js" : "bundle.min.js";
```
I think we need to update `src/platforms/javascript/index.mdx` and `src/wizard/javascript/index.md`, they reference `bundle.apm.min.js`.

rhcarvalho replied:
Also `src/wizard/javascript/react.md`.

rhcarvalho reviewed (changes requested):
Marking here that some references are still pointing to `@sentry/apm`.

```
$ git grep --line-number apm
src/platforms/javascript/index.mdx:136:  src="https://browser.sentry-cdn.com/{{ packages.version('sentry.javascript.browser') }}/bundle.apm.min.js"
src/platforms/javascript/index.mdx:137:  integrity="sha384-{{ packages.checksum('sentry.javascript.browser', 'bundle.apm.min.js', 'sha384-base64') }}"
src/wizard/javascript/index.md:12:  src="https://browser.sentry-cdn.com/{{ packages.version('sentry.javascript.browser') }}/bundle.apm.min.js"
src/wizard/javascript/index.md:13:  integrity="sha384-{{ packages.checksum('sentry.javascript.browser', 'bundle.apm.min.js', 'sha384-base64') }}"
src/wizard/javascript/index.md:22:import { Integrations as ApmIntegrations } from '@sentry/apm';
src/wizard/javascript/react.md:17:$ yarn add @sentry/react @sentry/apm
src/wizard/javascript/react.md:20:$ npm install @sentry/react @sentry/apm
src/wizard/javascript/react.md:33:import { Integrations } from '@sentry/apm';
```

AbhiPrasad commented:
@rhcarvalho I removed all references to `apm`. Can you please take another look at this PR? 

Also, I kept `import { Integrations } from "@sentry/tracing";` because we combine `Integrations` in the CDN bundle. To keep behaviour consistent, I changed this. We can come back to review this when we look at integrations in general for v6.

rhcarvalho commented on src/docs/product/performance/apm-to-tracing.mdx:
```diff
@@ -0,0 +1,107 @@
+
+## Migrating from @sentry/apm to @sentry/tracing
+
+For most cases, just replacing your `@sentry/tracing` import is enough to switch packages.
```
Looking at the examples, it seems that more realistically at least 2 changes are required (for the majority of users):

```diff
 -import "@sentry/apm";
 +import "@sentry/tracing";
 ```

and

```diff
 Sentry.init({
    integrations: [
 -    new Integrations.Tracing(),
 +    new Integrations.BrowserTracing(),
    ]
  });
```


Suggestion:

Break this down into `Browser` and `Node`. For browser the change is as above or the new CDN snippet (replacing the `bundle.apm.min.js` one).

For Node.js, is the import change enough?

AbhiPrasad replied:
Yes for Node the import change is enough. I guess if you were using Express, there is an added step. I will change.

AbhiPrasad replied:
Wait, actually we should be fine because we didn't change the name for Express.

rhcarvalho commented on src/docs/product/performance/apm-to-tracing.mdx:
````diff
@@ -0,0 +1,107 @@
+ });
+```
+
+If you were using the `beforeNavigate` option from the `Tracing` integration, the API for the `beforeNavigate` option in `BrowserTracing` has changed slightly. Instead of passing in a location and returning a string representing transaction name, `beforeNavigate` now accepts a transaction context and is expected to return transaction context. You can now add extra tags or change the op based on d ...
````
I'm missing a mention here about the rename `beforeNavigation` -> `beforeNavigate` (or was it just the docs that were wrong before?)

AbhiPrasad replied:
Docs were wrong before, it was always called `beforeNavigate`.

rhcarvalho replied:
I think bad docs before.

rhcarvalho commented on src/includes/performance-monitoring/configuration/javascript.mdx:
```diff
@@ -1,36 +1,42 @@
 
+<Alert level="warning" title="Note"><markdown>
+
+We switched our package name from `@sentry/apm` to `@sentry/tracing`. If you were previously using `@sentry/apm`, please see [Switching to the tracing package](/product/performance/apm-to-tracing/)
```
I'm going to apply Fiona's suggestion to match the page title.

rhcarvalho commented on src/includes/performance-monitoring/configuration/javascript.mdx:
```diff
@@ -1,36 +1,42 @@
 
+<Alert level="warning" title="Note"><markdown>
+
+We switched our package name from `@sentry/apm` to `@sentry/tracing`. If you were previously using `@sentry/apm`, please see [Switching to @sentry/tracing](/product/performance/apm-to-tracing/)
```
```suggestion
We switched our package name from `@sentry/apm` to `@sentry/tracing`. If you were previously using `@sentry/apm`, please see [Switching to @sentry/tracing](/product/performance/apm-to-tracing/).
```

rhcarvalho commented on src/includes/performance-monitoring/configuration/react.mdx:
```diff
@@ -32,7 +38,7 @@ ReactDOM.render(<App />, rootNode);
 **`withProfiler` Higher-Order Component**
 
-`@sentry/react` exports a `withProfiler` higher-order component that leverages the `@sentry/apm` Tracing integration to add React related spans to transactions. If the Tracing integration is not enabled, the Profiler will not work.
+`@sentry/react` exports a `withProfiler` higher-order component that leverages the `@sentry/tracing` to add React-related spans to transactions.
```
```suggestion
`@sentry/react` exports a `withProfiler` higher-order component that leverages the `@sentry/tracing` package to add React-related spans to transactions.
```

rhcarvalho commented on src/includes/performance-monitoring/configuration/vue.mdx:
```diff
@@ -4,39 +4,45 @@ To get started with performance monitoring with Vue.js, first install these pack
 Sentry injects a handler inside Vue's `beforeCreate` mixin, providing access to a Vue component during its life cycle stages.
-When Sentry encounters a component named `root`, which is a top-level Vue instance (as in `new Vue({})`), we use our AM Tracing integration,
-and create a new activity named `Vue Application Render`. Once the activity has been created, it will wait until all of its child components render, and there aren't new rendering events triggered within the configured `timeout`, before marking the activity as completed.
+When Sentry encounters a component named `root`, which is a top-level Vue instance (as in `new Vue({})`), we use our AM BrowserTracing integration,
```
> and there aren't new rendering events triggered within the configured `timeout`

Did this behavior change?

AbhiPrasad replied:
I changed this because I felt that we didn't need to expose activities/`idleTransaction` concepts here.

If we really want to do this, we can do it in `BrowserTracing`.

In addition, with `@sentry/tracing`, the Vue integration now can work without `BrowserTracing` and `idleTransactions`, so referring to activities/`timeout` is not always correct.

rhcarvalho reviewed (approved):
💯
